### ISSUE TITLE
mkosi.finalize : $BUILDROOT not working

### ISSUE THREAD

**Issue by dud225** [systemd/mkosi#316]
Hello

the $BUILDROOT mecanism doesn't work as publicised : 
1) in a new build, the $BUILDROOT directory exists but is empty. Putting files there doesn't make them appear in the final image
2) in an incremental build, the $BUILDROOT directory doesn't even exist

How to reproduce ?
Script `mkosi.finalize` : 
```sh
#!/bin/bash
set -x
echo "env :"
env
touch ${BUILDROOT}/test_finalize_${1}
ls ${BUILDROOT}
find ${BUILDROOT}
```

Output of `mkosi -iff` : 
```sh
[...]
‣ Successfully installed cache copy /home/user/.local/share/libvirt/images/dev-vm.cache-pre-inst.
‣ Running finalize script...
+ echo 'env :'
env :
+ env
[...]
BUILDROOT=/var/tmp/mkosi-2lk5bohb/root
[...]
+ touch /var/tmp/mkosi-2lk5bohb/root/test_finalize_build
+ ls /var/tmp/mkosi-2lk5bohb/root
test_finalize_build
+ find /var/tmp/mkosi-2lk5bohb/root
/var/tmp/mkosi-2lk5bohb/root
/var/tmp/mkosi-2lk5bohb/root/test_finalize_build
‣ Running finalize script complete.
‣ Running finalize script...
+ echo 'env :'
env :
+ env
[...]
BUILDROOT=/var/tmp/mkosi-2lk5bohb/root
[...]
+ touch /var/tmp/mkosi-2lk5bohb/root/test_finalize_final
+ ls /var/tmp/mkosi-2lk5bohb/root
test_finalize_build  test_finalize_final
+ find /var/tmp/mkosi-2lk5bohb/root
/var/tmp/mkosi-2lk5bohb/root
/var/tmp/mkosi-2lk5bohb/root/test_finalize_build
/var/tmp/mkosi-2lk5bohb/root/test_finalize_final
‣ Running finalize script complete.
[...]
```
Now booting the VM and looking for these files : 
```sh
root@localhost:~# ls /
bin   efi   initrd.img      lib32   lost+found  opt   run   sys  var
boot  etc   initrd.img.old  lib64   media       proc  sbin  tmp  vmlinuz
dev   home  lib             libx32  mnt         root  srv   usr  vmlinuz.old
```

Now running mkosi in incremental mode `mkosi -if` : 
```sh
[...]
‣ Running finalize script...
+ echo 'env :'
env :
+ env
[...]
BUILDROOT=/var/tmp/mkosi-44iuy2pl/root
[...]
+ touch /var/tmp/mkosi-44iuy2pl/root/test_finalize_build
touch: cannot touch '/var/tmp/mkosi-44iuy2pl/root/test_finalize_build': No such file or directory
+ ls /var/tmp/mkosi-44iuy2pl/root
ls: cannot access '/var/tmp/mkosi-44iuy2pl/root': No such file or directory
+ find /var/tmp/mkosi-44iuy2pl/root
find: ‘/var/tmp/mkosi-44iuy2pl/root’: No such file or directory
Traceback (most recent call last):
  File "/usr/local/bin/mkosi", line 4486, in <module>
    main()
  File "/usr/local/bin/mkosi", line 4482, in main
    run_verb(args)
  File "/usr/local/bin/mkosi", line 4452, in run_verb
    build_stuff(args)
  File "/usr/local/bin/mkosi", line 4269, in build_stuff
    run_finalize_script(args, workspace.name, verb='build')
  File "/usr/local/bin/mkosi", line 2050, in run_finalize_script
    run([args.finalize_script, verb], env=env, check=True)
  File "/usr/local/bin/mkosi", line 69, in run
    return subprocess.run(cmdline, **kwargs)
  File "/usr/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/home/user/tools/mkosi/dev-ct/mkosi.finalize', 'build']' returned non-zero exit status 1.
```

**Comment by lucasdemarchi** [comment]
Currently it's `mkosi.build` that is supposed to put files in a BUILDROOT. Maybe we could also run it for `mkosi.finalize`

**Comment by dud225** [comment]
Here is an excerpt from the manual regarding `mkosi.finalize` :
> The environment variable $BUILDROOT points to the root directory of the installation image.
> [...]
> This script is more flexible than mkosi.postinst in two regards: it has access to the host file system so it's easier to copy in additional files or to modify the image based on external configuration

**Comment by lucasdemarchi** [comment]
Yeah... But you didn't have a build script to populate the BUILDROOT... Does it work if you have one?

**Comment by dud225** [comment]
erf unable to test, the current master branch fails to even run the build script
```sh
‣ Running build script...
Failed to change to specified working directory /root/src: No such file or directory
Traceback (most recent call last):
  File "/usr/local/bin/mkosi", line 4486, in <module>
    main()
  File "/usr/local/bin/mkosi", line 4482, in main
    run_verb(args)
  File "/usr/local/bin/mkosi", line 4452, in run_verb
    build_stuff(args)
  File "/usr/local/bin/mkosi", line 4275, in build_stuff
    run_build_script(args, workspace.name, raw)
  File "/usr/local/bin/mkosi", line 4196, in run_build_script
    run(cmdline, check=True)
  File "/usr/local/bin/mkosi", line 69, in run
    return subprocess.run(cmdline, **kwargs)
  File "/usr/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['systemd-nspawn', '--quiet', '--image=/home/user/.local/share/libvirt/images/.mkosi-k18bt9up', '--uuid=67a1c3a61beb4ad8b196bb3f0373fd1e', '--machine=mkosi-f8757882756e4d5db9b9189ec2242ad5', '--as-pid2', '--register=no', '--bind', '/var/tmp/mkosi-ygxj_cxv/dest:/root/dest', '--bind=/var/tmp/mkosi-ygxj_cxv/var-tmp:/var/tmp', '--setenv=WITH_DOCS=0', '--setenv=WITH_TESTS=1', '--setenv=WITH_NETWORK=0', '--setenv=DESTDIR=/root/dest', '--setenv=SRCDIR=/root/src', '--chdir=/root/src', '--private-network', '/root/mkosi.build']' returned non-zero exit status 1.
```

**Comment by nergdron** [comment]
I've just encountered what seems to be related, which is that mkosi.finalize isn't running in the right place in the workflow, at least as described by the docs:

"mkosi.finalize may be an executable script. If it exists it is invoked as last step of preparing an image, from the host system. It is once called for the development image (if this is enabled, see above) with the "build" command line parameter, as the last step before invoking the build script, after the mkosi.postinst script is invoked. It is called the second time with the "final" command line parameter as the last step before the image is considered complete. The environment variable $BUILDROOT points to the root directory of the installation image."

however, when I make a finalize script that does "echo FINAL1" when `$1 == build` and "echo FINAL2" when `$1 == final`, I see that it's attempting to run before the image is even partitioned, let alone after all mkosi.postinst. Note that mkosi.postinst does seem to run as documented, after the rest of the image setup has completed, including os and package installation.

```
‣ Temporary workspace in /var/tmp/mkosi-1nghek1_ is now set up.
‣ Running finalize script...
FINAL1
‣ Running finalize script complete.
‣ Running finalize script...
FINAL2
‣ Running finalize script complete.
‣ Creating partition table...
Checking that no-one is using this disk right now ... OK
```

if I echo "${BUILDROOT}" from mkosi.finalize, it shows "/var/tmp/mkosi-1nghek1_/root", which doesn't exist at this stage in the build, so I'm not sure how mkosi.finalize can work as documented. it certainly isn't running after mkosi.postinst like it's supposed to.

**Comment by nergdron** [comment]
also in testing I tried creating a mkosi.build script that just contains `echo BUILD` to see if that made a difference, and I get the same stack trace and reference to /root/src as mentioned back in january, so it seems like mkosi.build is still broken as well.

**Comment by dellgreen** [comment]
seeing the same problem still

**Comment by DaanDeMeyer** [comment]
Should be fixed by #432
